lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+		duration := time.Since(timeBeforeSending)
+		buffer := time.Duration(1) * time.Millisecond
+		var nextDelay time.Duration
+		switch {
+		case duration < w.notifyDelay/10:
+			nextDelay = w.notifyDelay
+		case duration+buffer > w.notifyDelay:
+			nextDelay = buffer
+		default:
+			nextDelay = w.notifyDelay - duration
+		}
+		select {
+		case w.notifyTimerResetChan <- nextDelay:
+		case <-w.stop:
+		}
+	}()
+	return
+}
+
+// popOldEvents removes events that should be sent to
+func (w *watcher) popOldEvents(dir *eventDir, currTime time.Time) eventBatch {
+	oldEvents := make(eventBatch)
+	for _, childDir := range dir.dirs {
+		for path, event := range w.popOldEvents(childDir, currTime) {
+			oldEvents[path] = event
+		}
+	}
+	for path, event := range dir.events {
+		// 2 * results in mean event age of notifyDelay (assuming randomly
+		// occurring events). Reoccuring and remove/mixed events
+		// (for efficient renames) events are delayed until notifyTimeout.
+		if (event.evType == nonRemove && 2*currTime.Sub(event.lastModTime) > w.notifyDelay) || currTime.Sub(event.firstModTime) > w.notifyTimeout {
+			oldEvents[path] = event
+			delete(dir.events, path)
+		}
+	}
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+	return oldEvents
+}
+
+func (w *watcher) updateInProgressSet(event events.Event) {
+	if event.Type == events.ItemStarted {
+		path := event.Data.(map[string]string)["item"]
+		w.inProgress[path] = struct{}{}
+	} else if event.Type == events.ItemFinished {
+		path := event.Data.(map[string]interface{})["item"].(string)
+		delete(w.inProgress, path)
+	}
+}
+
+func (w *watcher) pathInProgress(path string) bool {
+	_, exists := w.inProgress[path]
+	return exists
+}
+
+func (w *watcher) UpdateIgnores(ignores *ignore.Matcher) {
+	l.Debugln(w, "Ignore patterns update")
+	select {
+	case w.folderIgnoresUpdate <- ignores:
+	case <-w.stop:
+	}
+}
+
+func (w *watcher) String() string {
+	return fmt.Sprintf("fswatcher/%s:", w.folderDescription)
+}
+
+func (w *watcher) eventType(notifyType notify.Event) eventType {
+	if notifyType&w.removeEventMask() != 0 {
+		return remove
+	}
+	return nonRemove
+}
+
+func (w *watcher) VerifyConfiguration(from, to config.Configuration) error {
+	return nil
+}
+
+func (w *watcher) CommitConfiguration(from, to config.Configuration) bool {
+	found := false
+	var cfg config.FolderConfiguration
+	for _, cfg = range to.Folders {
+		if cfg.ID == w.folderID {
+			found = true
+		}
+	}
+	if !found {
+		// Nothing to do, model will soon stop this service
+		return true
+	}
+	select {
+	case w.configUpdate <- cfg:
+	case <-w.stop:
+	}
+	return true
+}
+
+func (w *watcher) updateConfig(cfg config.FolderConfiguration) {
+	w.folderPath = filepath.Clean(cfg.Path())
+	w.folderDescription = cfg.Description()
+	w.notifyDelay = time.Duration(cfg.FSWatcherDelayS) * time.Second
+	w.notifyTimeout = notifyTimeout(cfg.FSWatcherDelayS)
+	w.folderIgnorePerms = cfg.IgnorePerms
```

Comment:
perhaps you should just store w.cfg and access w.cfg.BlahBlah, making it clear where the values come from.